src/Evolution/Particles/MonteCarlo/EvolveMonteCarloPackets.cpp

```diff
@@ -0,0 +1,127 @@
+    const Scalar<DataVector>& lapse,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& shift,
+    const tnsr::i<DataVector, 3, Frame::Inertial>& d_lapse,
+    const tnsr::iJ<DataVector, 3, Frame::Inertial>& d_shift,
+    const tnsr::iJJ<DataVector, 3, Frame::Inertial>& d_inv_spatial_metric,
+    const tnsr::II<DataVector, 3, Frame::Inertial>& inv_spatial_metric,
+    const size_t& closest_point_index) {
+  // Temporary variables
+  std::vector<double> dpdt(3, 0);
+  std::vector<double> dxdt(3, 0);
+  std::vector<double> p0(3, 0);
+
+  // Calculate p^t from normalization of 4-momentum
+  packet->renormalize_momentum(inv_spatial_metric, lapse, closest_point_index);
+
+  // Calculate time derivative of 3-momentum at beginning of time step
+  detail::time_derivative_momentum_geodesic(&dpdt, *packet, lapse, d_lapse,
+                                            d_shift, d_inv_spatial_metric,
+                                            closest_point_index);
+  // Take half-step (momentum only, as time derivative is independent of
+  // position)
+  for (size_t i = 0; i < 3; i++) {
+    // Store momentum at beginning of step
+    p0[i] = packet->momentum[i];
+    packet->momentum[i] += dpdt[i] * time_step * 0.5;
+  }
+  // Calculate p^t from normalization of 4-momentum
+  packet->renormalize_momentum(inv_spatial_metric, lapse, closest_point_index);
+
+  // Calculate time derivative of 3-momentum and position at half-step
+  detail::time_derivative_momentum_geodesic(&dpdt, *packet, lapse, d_lapse,
+                                            d_shift, d_inv_spatial_metric,
+                                            closest_point_index);
+  for (size_t i = 0; i < 3; i++) {
+    dxdt[i] = (-1.0) * shift.get(i)[closest_point_index];
+    for (size_t j = 0; j < 3; j++) {
+      dxdt[i] += inv_spatial_metric.get(i, j)[closest_point_index] *
+                 packet->momentum[j] / packet->momentum_upper_t;
+    }
+  }
+  // Take full time step
+  for (size_t i = 0; i < 3; i++) {
+    packet->momentum[i] = p0[i] + dpdt[i] * time_step;
+    packet->coordinates[i] += dxdt[i] * time_step;
+  }
+  packet->time += time_step;
+}
+
+}  // namespace detail
+
+void evolve_packets(
+    const gsl::not_null<std::vector<Packet>*> packets, const double& time_step,
+    const Mesh<3>& mesh,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& mesh_coordinates,
+    const Scalar<DataVector>& lapse,
+    const tnsr::I<DataVector, 3, Frame::Inertial>& shift,
+    const tnsr::i<DataVector, 3, Frame::Inertial>& d_lapse,
+    const tnsr::iJ<DataVector, 3, Frame::Inertial>& d_shift,
+    const tnsr::iJJ<DataVector, 3, Frame::Inertial>& d_inv_spatial_metric,
+    const tnsr::II<DataVector, 3, Frame::Inertial>& inv_spatial_metric) {
+  // Mesh information. Currently assumes uniform grid without map
+  const Index<3>& extents = mesh.extents();
+  const std::vector<double> bottom_coord_mesh{mesh_coordinates.get(0)[0],
```

Comment:
Can you please change this and the next `std::vector` to `std::array<double, 3>`?